lib/ansible/modules/cloud/docker/docker_swarm_service.py

```diff
@@ -0,0 +1,902 @@
+        if self.restart_policy!=os.restart_policy:
+            differences.append('restart_policy')
+        if self.restart_policy_attempts!=os.restart_policy_attempts:
+            differences.append('restart_policy_attempts')
+        if self.restart_policy_delay!=os.restart_policy_delay:
+            differences.append('restart_policy_delay')
+        if self.restart_policy_window!=os.restart_policy_window:
+            differences.append('restart_policy_window')
+        if self.image!=os.image:
+            differences.append('image')
+        if self.user!=os.user:
+            differences.append('user')
+        if self.dns!=os.dns:
+            differences.append('dns')
+        if self.dns_search!=os.dns_search:
+            differences.append('dns_search')
+        if self.dns_options!=os.dns_options:
+            differences.append('dns_options')
+        if self.hostname!=os.hostname:
+            differences.append('hostname')
+        if self.tty!=os.tty:
+            differences.append('tty')
+        return len(differences)>0, differences, needs_rebuild
+
+    def __str__(self):
+        return str({
+            'mode': self.mode,
+            'env': self.env,
+            'endpoint_mode': self.endpoint_mode,
+            'mounts': self.mounts,
+            'networks': self.networks,
+            'replicas': self.replicas})
+    def generate_docker_py_service_description(self, name, docker_networks):
+        cspec={'Image': self.image,
+               'User': self.user}
+        cspec['Mounts']=[]
+        for mount_config in self.mounts:
+            cspec['Mounts'].append({
+                'Target': mount_config['target'],
+                'Source': mount_config['source'],
+                'Type': mount_config['type'],
+                'ReadOnly': mount_config['readonly']
+            })
+        cspec['DNSConfig']={"Nameservers": self.dns,
+                            "Search": self.dns_search,
+                            "Options": self.dns_options}
+        cspec['Args']=self.args
+        cspec['Env']=self.env
+        cspec['TTY']=self.tty
+        cspec['Hostname']=self.hostname
+        cspec['Labels']=self.container_labels
+        restart_policy=types.RestartPolicy(
+            condition     = self.restart_policy,
+            delay         = self.restart_policy_delay,
+            max_attempts  = self.restart_policy_attempts,
+            window        = self.restart_policy_window)
+        resources={
+            'Limits': {
+                'NanoCPUs': int(self.limit_cpu * 1000000000),
+                'MemoryBytes': self.limit_memory*1024*1024},
+            'Reservations': {
+                'NanoCPUs': int(self.reserve_cpu * 1000000000),
+                'MemoryBytes': self.reserve_memory*1024*1024}}
+        task_template=types.TaskTemplate(
+            container_spec  = cspec,
+            restart_policy  = restart_policy,
+            placement       = self.constraints,
```

Comment:
The conversion you suggest is handled in [the docker python library](https://github.com/docker/docker-py/blob/master/docker/types/services.py#L35), and checking the older versions docs it looks like the `placement` parameter has not changed between 1.24, 1.25 and 1.26.

Can you show me the module call which gave you problems (and the error you've received)?

I tested this parameter on docker 1.12.6, 1.30.0, and 1.30.1 with api_version 1.24, 1.25 and 1.26 and got no problems; this is the syntax I used:
```
docker_swarm_service:
    name: myservice
    image: "nginx:1.10.2-alpine@sha256:56766e6df28bb0c572dca24acbe5639ddf866d931c105cb33410d52624c27682"
    constraints:
    - "node.labels.type==ok"
```